### PR TITLE
Add signal timeout for 'stop_workers'

### PR DESCRIPTION
**Pull request by BalaBalaYi**
### What changes were proposed in this pull request?

Will raise error(pod failover) if the 'stop_workers' timeout.

### Why are the changes needed?

'stop_workers' may hang if there is hardware issue.

### Does this PR introduce any user-facing change?

No.

### How was this patch tested?

UT.

### CODE REVIEW

File: dlrover/python/elastic_agent/torch/training.py [hunk 1]
```diff
@@ -571,12 +571,26 @@ def _initialize_workers(self, worker_group):
...
+
+            signal.alarm(0)
+        except TimeoutError as te:
+            logger.error(str(te))
+            raise
+        finally:
+            signal.alarm(0)
+
+    def _stop_timeout_handler(self, signum, frame):
+        raise TimeoutError("Timed out waiting for stopping workers.")
```
Comment (majieyue) [inline dlrover/python/elastic_agent/torch/training.py:593]: 这里加点日志吧

File: dlrover/python/elastic_agent/torch/training.py [hunk 2]
```diff
@@ -571,12 +571,26 @@ def _initialize_workers(self, worker_group):
...
+        try:
+            signal.signal(signal.SIGALRM, self._stop_timeout_handler)
+            signal.alarm(timeout)
+
+            if self._config.accelerator == Accelerators.ASCEND_NPU:
+                logger.info("stop workers via SIGKILL")
+                self._shutdown(death_sig=signal.SIGKILL)
+            else:
+                super()._stop_workers(worker_group)
+
```
Comment (majieyue) [inline dlrover/python/elastic_agent/torch/training.py:584]: 这里也加日志

File: dlrover/python/tests/test_elastic_training_agent.py [hunk 3]
```diff
@@ -254,8 +258,8 @@ def setUp(self) -> None:
             rdzv_handler=self.rdzv_handler,
             max_restarts=self.config.max_restarts,
             monitor_interval=self.config.monitor_interval,
-            redirects=self.config.redirects,
-            tee=self.config.tee,
+            # redirects=self.config.redirects,
+            # tee=self.config.tee,
```
Comment (majieyue) [inline dlrover/python/tests/test_elastic_training_agent.py:262]: 这里为啥要注释掉？

File: dlrover/python/elastic_agent/torch/training.py (hunk 1 above)
Comment (BalaBalaYi) [reply dlrover/python/elastic_agent/torch/training.py:593]: log here:
image

**Review by majieyue — APPROVED** [review]

File: dlrover/python/elastic_agent/torch/training.py (hunk 2 above)
Comment (BalaBalaYi) [reply dlrover/python/elastic_agent/torch/training.py:585]: There is already enough info to judge whether is hanged here.

**Comment by majieyue** [comment]
lGTM

**Pull request merged**